### PR TITLE
Split rFXGen implementation into its own header

### PR DESCRIPTION
**Pull request by RobLoach**
Splitting the WaveParams-specific API to its own `RFXGEN_IMPLEMENTATION` `rfxgen.h` header allows other raylib applications to integrate with its API to generate new wave sounds on the fly.

### CODE REVIEW

File: src/rfxgen.c [hunk 1]
```diff
@@ -143,69 +147,16 @@ bool __stdcall FreeConsole(void);       // Close console from code (kernel32.lib
 
 // Simple log system to avoid printf() calls if required
 // NOTE: Avoiding those calls, also avoids const strings memory usage
-#define SUPPORT_LOG_INFO
```
Comment (RobLoach) [inline src/rfxgen.c:146]:
``` suggestion
#define SUPPORT_LOG_INFO
```

File: src/rfxgen.c [hunk 2]
```diff
@@ -245,25 +195,6 @@ static void ShowCommandLineInfo(void);                      // Show command line
...
-static WaveParams GenExplosion(void);       // Generate sound: Explosion
-static WaveParams GenPowerup(void);         // Generate sound: Powerup
-static WaveParams GenHitHurt(void);         // Generate sound: Hit/Hurt
-static WaveParams GenJump(void);            // Generate sound: Jump
-static WaveParams GenBlipSelect(void);      // Generate sound: Blip/Select
-static WaveParams GenRandomize(void);       // Generate random sound
-static void WaveMutate(WaveParams *params); // Mutate current sound
-
-// Auxiliar functions
-static void DrawWave(Wave *wave, Rectangle bounds, Color color);    // Draw wave data using lines
```
Comment (raysan5) [inline src/rfxgen.c:266]: I think that function should stay in the main problem. `rfxgen.h` should be as much independent of raylib as possible.

File: src/rfxgen.c [hunk 3]
```diff
@@ -134,6 +134,10 @@
     #include <fcntl.h>
 #endif
 
+#define RFXGEN_IMPLEMENTATION
+#define RFXGEN_LOG(...) printf(__VA_ARGS__)
```
Comment (raysan5) [inline src/rfxgen.c:138]: This line can just be included in `rfxgen.h`

**Comment by raysan5** [comment]
@RobLoach This is a nice improvement that I've been actually considering for years (to expose wave generation as a single-file header-only library) but I never implemented. Thanks for the push on that direction! 

I added some review comments, I also think `rfxgen.h` still needs some work. It would be great if it could be completely independant of raylib. I can review it more carefully later on.

File: src/rfxgen.c [hunk 4]
```diff
@@ -1172,842 +1103,6 @@ static void ProcessCommandLine(int argc, char *argv[])
...
 
-//--------------------------------------------------------------------------------------------
-// Load/Save/Export functions
-//--------------------------------------------------------------------------------------------
-
-// Reset wave parameters
-static void ResetWaveParams(WaveParams *params)
-{
-    // NOTE: Random seed is set to a random value
-    params->randSeed = GetRandomValue(0x1, 0xFFFE);
```
Comment (raysan5) [inline src/rfxgen.c:1183]: Avoid raylib func: `GetRandomValue()` can be replaced by `rand()`

File: src/rfxgen.c [hunk 5]
```diff
@@ -1172,842 +1103,6 @@ static void ProcessCommandLine(int argc, char *argv[])
...
-    params->lpfCutoffValue = 1.0f;
-    params->lpfCutoffSweepValue = 0.0f;
-    params->lpfResonanceValue = 0.0f;
-    params->hpfCutoffValue = 0.0f;
-    params->hpfCutoffSweepValue = 0.0f;
-}
-
-// Generates new wave from wave parameters
-// NOTE: By default wave is generated as 44100Hz, 32bit float, mono
-static Wave GenerateWave(WaveParams params)
```
Comment (raysan5) [inline src/rfxgen.c:1229]: Avoid raylib type: `Wave` can be replaced by custom `rfxWave` and mapping done on user side.

File: src/rfxgen.c [hunk 6]
```diff
@@ -1172,842 +1103,6 @@ static void ProcessCommandLine(int argc, char *argv[])
...
-
-    Wave genWave = { 0 };
-    genWave.frameCount = sampleCount/1;    // Number of samples / channels
-    genWave.sampleRate = WAVE_SAMPLE_RATE; // By default 44100 Hz
-    genWave.sampleSize = 32;               // By default 32 bit float samples
-    genWave.channels = 1;                  // By default 1 channel (mono)
-
-    // NOTE: Wave can be converted to desired format after generation
-
-    genWave.data = (float *)RL_CALLOC(genWave.frameCount*genWave.channels, sizeof(float));
```
Comment (raysan5) [inline src/rfxgen.c:1525]: `rfxgen.h` can implement its own `RFXGEN_CALLOC` and `RFXGEN_FREE`

File: src/rfxgen.c [hunk 7]
```diff
@@ -1172,842 +1103,6 @@ static void ProcessCommandLine(int argc, char *argv[])
...
-
-    return genWave;
-}
-
-// Load .rfx (rFXGen) or .sfs (sfxr) sound parameters file
-static WaveParams LoadWaveParams(const char *fileName)
-{
-    WaveParams params = { 0 };
-
-    if (IsFileExtension(fileName, ".rfx"))
```
Comment (raysan5) [inline src/rfxgen.c:1538]: This function should probably be re-implemented internally in the module.

File: src/rfxgen.c [hunk 8]
```diff
@@ -1172,842 +1103,6 @@ static void ProcessCommandLine(int argc, char *argv[])
...
-                fread(&length, 1, sizeof(unsigned short), rfxFile);
-
-                if (version != 200) LOG("[%s] rFX file version not supported (%i)\n", fileName, version);
-                else
-                {
-                    if (length != sizeof(WaveParams)) LOG("[%s] Wrong rFX wave parameters size\n", fileName);
-                    else fread(&params, 1, sizeof(WaveParams), rfxFile);   // Load wave generation parameters
-                }
-            }
-            else LOG("[%s] rFX file does not seem to be valid\n", fileName);
```
Comment (raysan5) [inline src/rfxgen.c:1566]: `RFXGEN_LOG()` should be used inside the module.

**Comment by RobLoach** [comment]
@raysan5 I've made the majority of updates based on your feedback. Thanks a lot!

Unsure of a good design pattern around the `Wave` struct though.

**Comment by raysan5** [comment]
@RobLoach Thank you very much for the reviews! I'm merging and we can continue from here! 
About Wave struct, an alternative version can be provided by rfxgen.h and the mapping can be done to raylib one on main program.

**Comment by raysan5** [comment]
Personally I think approach 2) is the best one. I propose a `struct rfxWave`.

**Pull request merged**